### PR TITLE
Add voucher tools to reverse and correct posted records

### LINKED ISSUE #3063: Design: make corrections of posted records easier by adding "reverse" and "correct" actions
The point of the unposted/posted dichotomy is to have correct data validated and verified for reports.  Unfortunately, bad data is still posted and needs to be corrected.  The proper way to correct the transaction is to reverse it, then recreate it with the correct values/accounts/currencies/etc.

When you actually try to correct a transaction, there are a lot of steps involved.  We should make this process easier.  I propose a simple(r) way to make changes to posted data, while still maintaining the paper trail, by either "reverse"-ing or "correct"-ing a previously made transaction.

**Design**
For unposted records, we provide the user the opportunity to "delete" records using a delete button on the transaction modal.  You can see that below:

deletebuttonontransactionedit
_Fig 1: Delete button on Transaction Edit Modal_

For posted records, this button is hidden.  I propose that we replace the "delete" button with a button group containing "correct" and "reverse" options.

Once a user clicks the "correct" button, the values in the transaction will be unlocked (as if it was unposted) for correction.  After the user makes adjustments, they will submit the modal.  When the user submits, the following things take place:
 1. The system creates a second, reversing transaction that annuls the first one.
 2. The system creates a third transaction based on the data in the transaction and the modifications the user had made in the edit modal.
 3. Both records will be written first as vouchers, then added to the journal via the `PostVoucher()` stored procedure.  They must be manually posted by the accountant to take effect as an additional verification step.

We would need heavy testing of the transaction edit modal for this to work effectively.  I'm not sure it is well tested at this point.

### LINKED ISSUE #3149: Proposal: `bh-voucher-tools` collection of inline tools
This issue is to track the implementation of the design found in https://github.com/IMA-WorldHealth/bhima-2.X/issues/3063. 

**Design specification:** 
https://github.com/IMA-WorldHealth/bhima-2.X/issues/3063 does a good idea of describing the feature set.

**Feature goals (in rough order of importance):** 
1. Do not further complicate the Transaction editing modal. This controller is already responsible for a lot of complicated features and including further functionality here would give it a lot of responsibilities. 
2. Reduce code maintenance surface area. A lot of the ideas encapsulated in `voucher-tools` already exist in client and server services and controllers respectively. 
3. Provide a uniform user experience for both the `reverse` and `correct` features. Reverse simply requires a single input value (transaction ID) and correct requires a complex set of transaction changes, however to the user they are presented as very similar concepts.

**Core concept:** 
* Tools that leverage creating vouchers to automate financial operations. 

**/client/src/modules/voucher-tools/README.md:** 

---
# Voucher Tools                                                                                                                                                                                                                                                                    
Voucher Tools is a name given to a collection of components and services that provide administrative tools for quickly creating vouchers based on different inputs.                                                                                                                
                                                                                                                                                                                                                                                                                   
## Structure                                                                                                                                                                                                                                                                       
The structure of this unit is to be decided. Currently each tool defines its own Component that is responsible for switching between input, output and error states.                                                                                                               
                                                                                                                                                                                                                                                                                   
## Tools                                                                                                                                                                                                                                                                           
### Reverse                                                                                                                                                                                                                                                                        
The Reverse tool allows reversal of any transaction in the posting journal or general ledger.                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                   
Input: Transaction to reverse                                                                                                                                                                                                                                                      
Process: Submit request to /journal/reverse to generate a voucher reversing this transaction                                                                                                                                                                                       
Output: Description of process and voucher used to reverse transaction                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                   
Example Usage:                                                                                                                                                                                                                                                                     
```html                                                                                                                                                                                                                                                                            
<bh-voucher-tools-reverse                                                                                                                                                                                                                                                          
  ng-if="Ctrl.isReversing"                                                                                                                                                                                                                                                         
  source="{ transaction_id : Ctrl.transaction_id }"                                                                                                                                                                                                                                
  show-badge>                                                                                                                                                                                                                                                                      
</bh-voucher-tools-reverse>                                                                                                                                                                                                                                                        
```                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                   
### Correct                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                   
Example Usage:                                                                                                                                                                                                                                                                     
```html                                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                   
<!-- Notice no show-badge is assigned, it is assumed that the Correct tool -->                                                                                                                                                                                                     
<!-- will be integrated in a page that will have to modify its own UI to -->                                                                                                                                                                                                       
<!-- provide the voucher rows required by Correct -->                                                                                                                                                                                                                              
<bh-voucher-tools-correct                                                                                                                                                                                                                                                          
  ng-if="Ctrl.isCorrecting"    
  source="{ transaction_id : Ctrl.transaction_id,   voucher_rows : Ctrl.rows }" />
</bh-voucher-tools-correct>                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                   
## Components                                                                                                                                                                                                                                                                      
```html                                                                                                                                                                                                                                                                            
<!-- An inline badge element to demonstrate that a voucher tool is active -->                                                                                                                                                                                                      
<!-- on the current module/ component -->                                                                                                                                                                                                                                          
<bh-voucher-tool-badge></bh-voucher-tool-badge>                                                                                                                                                                                                                                    
```                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                   
## Proposed Unify API                                                                                                                                                                                                                                                              
An API proposal to unify all voucher tools into an umbrella component, this would allow developers to simply include `<bh-voucher-tool>` and specify which functionality they require. This API should only be implemented when it is clear what is required for building a vouche>
                                                                                                                                                                                                                                                                                   
```html                                                                                                                                                                                                                                                                            
<!-- Example usage -->                                                                                                                                                                                                                                                             
<bh-voucher-tool                                                                                                                                                                                                                                                                   
  ng-if="Ctrl.isReversing"                                                                                                                                                                                                                                                         
  reverse-transaction />                                                                                                                                                                                                                                                           
</bh-voucher-tool>                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                   
<bh-voucher-tool                                                                                                                                                                                                                                                                   
  ng-if="Ctrl.isCorrecting"                                                                                                                                                                                                                                                        
  correct-transaction />
```
---

**Potential Alternatives:** 
* These tools could be developed as individual application modules that accept a parameter for example /voucher-tools/reverse/TPA1. This would give maximum screen real estate 
* Tools implemented as Modal ecosystem (similar to receipts)

### PR DESCRIPTION
**Pull request by jniles**
Documented thoroughly in discussion #3063 .

Closes #3063.  Closes #3149.

### CODE REVIEW

**Comment by jniles** [comment]
@mbayopanda, can you review the French translations for this?

**Comment by jniles** [comment]
@sfount, great work on this!  I've done manual tests and it works according to spec.  Well done!

I'm a bit nervous about merging ~2.5K lines of code without any tests.  Would you have a chance to write a few?

**Comment by jniles** [comment]
This seems bullet proof from my testing of it.  With the added translations, it's almost ready to go.  We need to do QA testing and add automated testing.  

Here's the plan for getting this in:

 - [x] Write integration tests for the added `/journal/:uuid/reverse` route.  This should thoroughly check that the transactions have been reversed.
 - [x] Write integration tests for [`/journal/:uuid/correct`](https://github.com/IMA-WorldHealth/bhima-2.X/pull/3206/files#diff-99de27cb2ded592bf6ae68b3d132eaceR215) route.
 - [ ] Write units tests for the added components, except those that are purely ui.  I'm not sure `bh-badge` needs one.

Once these conditions are met, we can merge this code.

**Comment by jniles** [comment]
@mbayopanda, this is ready for your review.

File: client/src/css/structure.css
Comment (mbayopanda) [inline client/src/css/structure.css:948]: css best practice : Rule must have all its properties in alphabetical order

File: client/src/css/structure.css
Comment (mbayopanda) [inline client/src/css/structure.css]: This rule is empty, is it possible to remove it

File: client/src/css/structure.css
Comment (mbayopanda) [inline client/src/css/structure.css]: css best practice : Values of 0 shouldn't have units specified

File: server/controllers/finance/journal/index.js
Comment (mbayopanda) [inline server/controllers/finance/journal/index.js]: Parenthesis are not necessary for one argument : 
```js
.then(reverseResult => VoucherService.lookupVoucher(reverseResult.uuid))
.then(voucher => res.status(201).json({ uuid : voucher.uuid, voucher }))
```

**Review by mbayopanda — COMMENTED** [review]
I made a first review and i just notice that when we correct a transaction with the voucher-tool, the transaction is submitted without values validation. When i set debit = 77 and credit = 75, i got : 

image

File: client/src/css/structure.css
Comment (jniles) [reply client/src/css/structure.css:948]: I like it. :+1:  Will do.

**Comment by jniles** [comment]
@mbayopanda, thanks for the testing and review.  Looks like I need to test:
 - [x] Error cases when something goes wrong in the correction process.
 - [ ] What happens with something has already been reversed and you try to edit it?  What is the anticipated behavior?

**Comment by jniles** [comment]
New behavior with offline validation:

_GIF 1: Offline validation on the records behaves the same way as the edit transaction modal_

**Comment by jniles** [comment]
@mbayopanda, this is ready for a second review. 👍

File: client/src/js/services/TransactionService.js
Comment (mbayopanda) [inline client/src/js/services/TransactionService.js:144]: This code looks good but i noticed that the validation is skipped for unbalanced amounts :

File: server/controllers/finance/journal/index.js
Comment (mbayopanda) [inline server/controllers/finance/journal/index.js:622]: 👍

File: client/src/modules/voucher-tools/correct/voucher-tools.correct.component.js
Comment (mbayopanda) [inline client/src/modules/voucher-tools/correct/voucher-tools.correct.component.js:18]: The entry `VOUCHERS.TOOLS.CORRECT.DESCRIPTION` is missing in the voucher.json file.

**Review by mbayopanda — APPROVED** [review]
This is a great work, well done

I made some comments but the main feature `correct` and `reverse` transaction works perfectly.

**Comment by jniles** [comment]
@mbayopanda, thanks for the review!  I'll look into the issues with validation.

File: client/src/js/services/TransactionService.js
Comment (jniles) [reply client/src/js/services/TransactionService.js:144]: Hmm... I can't reproduce this on chrome or firefox.  Let me see if I can reproduce it on a VM.

**Comment by mbayopanda** [comment]
@jniles the code works fine and validation too

**Comment by jniles** [comment]
Great work @sfount.  Thanks so much for your contribution!

**Pull request merged**
